libs/agno/agno/db/postgres/postgres.py

```diff
@@ -1162,6 +1189,210 @@ def delete_user_memory(self, memory_id: str) -> bool:
+            "updated_at": int(time.time()),
+            **metrics,
+        }
+
+    def _fetch_all_sessions_data(self, dates_to_process: list[date]) -> Optional[dict]:
+        """Return all session data for the given dates, for all session types.
+
+        Returns:
+            dict: A dictionary with dates as keys and session data as values, for all session types.
+
+        Example:
+        {
+            "2000-01-01": {
+                "agent": [<session1>, <session2>, ...],
+                "team": [...],
+                "workflow": [...],
+            }
+        }
+        """
+        if not dates_to_process:
+            return None
+
+        start_timestamp = int(datetime.combine(dates_to_process[0], datetime.min.time()).timestamp())
+        end_timestamp = int(datetime.combine(dates_to_process[-1] + timedelta(days=1), datetime.min.time()).timestamp())
+
+        all_sessions_data = {
+            date_to_process.isoformat(): {"agent": [], "team": [], "workflow": []}
+            for date_to_process in dates_to_process
+        }
+
+        session_types = [
+            (SessionType.AGENT, "agent", self.agent_session_table_name),
+            (SessionType.TEAM, "team", self.team_session_table_name),
+            (SessionType.WORKFLOW, "workflow", self.workflow_session_table_name),
+        ]
+
+        for session_type, key, table_name in session_types:
+            if not table_name:
+                continue
+            try:
+                sessions, _ = self.get_sessions_raw(
+                    session_type=session_type, start_timestamp=start_timestamp, end_timestamp=end_timestamp
+                )
+                for session in sessions:
+                    session_date = date.fromtimestamp(session.get("created_at", start_timestamp)).isoformat()
+                    if session_date in all_sessions_data:
+                        all_sessions_data[session_date][key].append(session)
+
+            except Exception as e:
+                log_debug(f"Could not fetch {key} sessions: {e}")
+
+        return all_sessions_data
+
+    def _get_dates_to_calculate_metrics_for(self, starting_date: date) -> list[date]:
+        """Return the list of dates to calculate metrics for."""
+        days_diff = (date.today() - starting_date).days + 1
+        if days_diff <= 0:
+            return []
+        return [starting_date + timedelta(days=x) for x in range(days_diff)]
+
+    def _get_metrics_calculation_starting_date(self, table: Table) -> Optional[date]:
+        """Get the first date for which metrics calculation is needed:
+
+        1. If there are metrics records, return the date of the first day without a complete metrics record.
+        2. If there are no metrics records, return the date of the first recorded session.
+        3. If there are no metrics records and no sessions records, return None.
+        """
+        with self.Session() as sess:
+            stmt = select(table).order_by(table.c.date.desc()).limit(1)
```

Comment:
Is there a chance that more than 1 is incomplete? I guess not?

Reply:
Correct, there should never be multiple incomplete rows